In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [35]:
visa_df=pd.read_csv(r"C:\Users\kisho\Documents\data set\Visadataset.csv")
cat=visa_df.select_dtypes(include="object").columns
num=visa_df.select_dtypes(exclude="object").columns
cat,num

(Index(['case_id', 'continent', 'education_of_employee', 'has_job_experience',
        'requires_job_training', 'region_of_employment', 'unit_of_wage',
        'full_time_position', 'case_status'],
       dtype='object'),
 Index(['no_of_employees', 'yr_of_estab', 'prevailing_wage'], dtype='object'))

**encoding**
- it means converting catagoricalcolumn data to numerical data
- because ml model expect the data should be in numerical format
- ml models devlop by math algorithm
- it is very very important to apply encoding
- the following methods are
- 1.map
- 2.np.where
- 3.one hot encoding
- 4.lable noder

**map**
- it is one method to convert catagorical value to numerical value.
- take one catagorical column
- get the unique lables first
- make a dictionary with assign a no. to each lable
- for example case status has 2 lables
- 1.cetified
- 2.denied
- assign 0 to certified ,1 to denied
- cret a dictionary lables as keys,no.asvalues only
- {certified:0,denied:1}

In [ ]:
#read the column
#get the unique lble
#make  a adictionary
#apply the map

In [3]:
visa_df['case_status'].unique()

array(['Denied', 'Certified'], dtype=object)

In [6]:
d={'Certified':0,'Denied':1}
visa_df['case_status_new']=visa_df['case_status'].map(d)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  
0          Denied                1  
1       Certified                0  
2          Denied                1  
3          Denied                1  
4       Certified                0  
...           ...              ...  
25475   Certified                0  
25476   Certified                0  
25477   Certified                0  
25478   Certified                0  
25479   Certified                0  

[25480 rows x 13 columns]

In [ ]:
#i want to apply the map method for all catagorical columns
#step_1:we need to get unique lbles
#step_3:we need to get a no.=len unique lables
#step_4:we need to creat a dictionary

In [19]:
unique_lables=sorted(visa_df['case_status'].unique())
list1=[i for i in range(len(unique_lables))] 
unique_lables,list1
d=dict(zip(unique_lables,list1))
visa_df['case_status']=visa_df['case_status'].map(d)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

In [22]:
for i in cat:
    unique_lables=sorted(visa_df[i].unique())
    list1=[i for i in range(len(unique_lables))]
    d=dict(zip(unique_lables,list1))
    visa_df[i]=visa_df[i].map(d)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**LableEncoder**

- it is a method to convert cat data to numerical data
- it is the same way how we did in the above
- the forloop map method we explored
- but the lable encoder also do the same
- it is under **scikit_learn** we usually called sklearn package
- under sklearn we a class called as preprocessing
- bcoz,currently we are ding data preprocessing only
- under preprocessing class we havea method called **LableEncoder**
- sklearn
- preprocessing
- LableEncoder
- any sklearn package has 3 steps
  - step_1:read the package
  - step_2:save the pacage
  - step_3:apply fit transform data

In [27]:
#step_1:read the package or load the method
from sklearn.preprocessing import LabelEncoder
#step_2:save the package/call the method
le=LabelEncoder()
#step_3:apply fit transform on data
visa_df['case_status']=le.fit_transform(visa_df['case_status'])
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

In [31]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
for i in cat:
    visa_df[i]=le.fit_transform(visa_df[i])
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**fit and transform**
- fit means devlop the application
- transform means process the logic
- if we observe in map method first wedevloped dictionary logic
- then we apply that dictionary to columns
- devloping the dictionary is akind of a fitting
- apply the dictionary to the olumn means we are traforming the data cat to numerical
- whenever we are transforming the data then use **fit transform**
- if we are not changing the data then only **fit**

**np.where**
- using np.where also we can change the cat data to numerical 
- but np.where is only for binary condition
- np.wher is same as if and else
- if condition is true: will apply true value
- if condition is false:will apply the false value
- so np.where use for only binary lables
- a column should have only two unique lables

In [ ]:
#value==certified replace with 0
#otherwise 1
np.where(con,0,1)

In [34]:
con=visa_df['case_status']=='Certified'
visa_df['case_status']=np.where(con,0,1)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

**one hot encoder**
- one hot means if one will be ON,another will be OFF
- ON represents with 1
- OFF represents with 0
- for example case status has 2 unique lables
1.certified
2.denied
- one hot encoder creats new columns which is equal to no.of unique lables
- for ex.case_status will creat two new extra columns
- case_status_certified
- case_status_denied

- in order to one hot encoder we have a method called
**pd.get_dummies**

**pd.get_dummies**

In [44]:
df=pd.get_dummies(visa_df['case_status'],dtype='int')
df['case_status']=visa_df['case_status'].values
df['dummy']=[i for i in range(len(visa_df))]
df

Certified  Denied case_status  dummy
0              0       1      Denied      0
1              1       0   Certified      1
2              0       1      Denied      2
3              0       1      Denied      3
4              1       0   Certified      4
...          ...     ...         ...    ...
25475          1       0   Certified  25475
25476          1       0   Certified  25476
25477          1       0   Certified  25477
25478          1       0   Certified  25478
25479          1       0   Certified  25479

[25480 rows x 4 columns]

In [ ]:
**join 2 data frame**

In [48]:
df1=pd.get_dummies(visa_df['case_status'],dtype='int')
df1

Certified  Denied
0              0       1
1              1       0
2              0       1
3              0       1
4              1       0
...          ...     ...
25475          1       0
25476          1       0
25477          1       0
25478          1       0
25479          1       0

[25480 rows x 2 columns]

In [46]:
df2=visa_df[['case_status']]
df2

case_status
0          Denied
1       Certified
2          Denied
3          Denied
4       Certified
...           ...
25475   Certified
25476   Certified
25477   Certified
25478   Certified
25479   Certified

[25480 rows x 1 columns]

**adding the 2 data frame**
- merge
- join
- concat
- append

In [49]:
pd.concat([df1,df2],axis=1)

Certified  Denied case_status
0              0       1      Denied
1              1       0   Certified
2              0       1      Denied
3              0       1      Denied
4              1       0   Certified
...          ...     ...         ...
25475          1       0   Certified
25476          1       0   Certified
25477          1       0   Certified
25478          1       0   Certified
25479          1       0   Certified

[25480 rows x 3 columns]

In [51]:
pd.get_dummies(visa_df,dtype="int")

no_of_employees  yr_of_estab  prevailing_wage  case_id_EZYV01  \
0                14513         2007         592.2029               1   
1                 2412         2002       83425.6500               0   
2                44444         2008      122996.8600               0   
3                   98         1897       83434.0300               0   
4                 1082         2005      149907.3900               0   
...                ...          ...              ...             ...   
25475             2601         2008       77092.5700               0   
25476             3274         2006      279174.7900               0   
25477             1121         1910      146298.8500               0   
25478             1918         1887       86154.7700               0   
25479             3195         1960       70876.9100               0   

       case_id_EZYV02  case_id_EZYV03  case_id_EZYV04  case_id_EZYV05  \
0                   0               0               0               0   
1                   1               0               0               0   
2                   0               1               0               0   
3                   0               0               1               0   
4                   0               0               0               1   
...               ...             ...             ...             ...   
25475               0               0               0               0   
25476               0               0               0               0   
25477               0               0               0               0   
25478               0               0               0               0   
25479               0               0               0               0   

       case_id_EZYV06  case_id_EZYV07  ...  region_of_employment_South  \
0                   0               0  ...                           0   
1                   0               0  ...                           0   
2                   0               0  ...                           0   
3                   0               0  ...                           0   
4                   0               0  ...                           1   
...               ...             ...  ...                         ...   
25475               0               0  ...                           1   
25476               0               0  ...                           0   
25477               0               0  ...                           1   
25478               0               0  ...                           0   
25479               0               0  ...                           0   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                              1                  1                   0   
1                              0                  0                   0   
2                              1                  0                   0   
3                              1                  0                   0   
4                              0                  0                   0   
...                          ...                ...                 ...   
25475                          0                  0                   0   
25476                          0                  0                   0   
25477                          0                  0                   0   
25478                          1                  0                   0   
25479                          0                  0                   0   

       unit_of_wage_Week  unit_of_wage_Year  full_time_position_N  \
0                      0                  0                     0   
1                      0                  1                     0   
2                      0                  1                     0   
3                      0                  1                     0   
4                      0                  1                     0   
...                  ...                ...                  

In [ ]:
- in one hot encoder a new column will created
- the new columns equal to no.of unique lables
- because of casse status two extra columns created
- because of continent 6 extra columns are created
- because of case id 25480 unique ids,25480 columns created


In [57]:
count=0
for i in cat:
    count=count+visa_df[i].nunique()
print(count+len(num))

25510


In [59]:
#except case id
count=0
for i in cat[1:]:
    count=count+visa_df[i].nunique()
print(count)
print(count+len(num))

27
30


In [61]:
for i in cat[1:]:
    print(visa_df[i].unique(),visa_df[i].nunique())


['Asia' 'Africa' 'North America' 'Europe' 'South America' 'Oceania'] 6
['High School' "Master's" "Bachelor's" 'Doctorate'] 4
['N' 'Y'] 2
['N' 'Y'] 2
['West' 'Northeast' 'South' 'Midwest' 'Island'] 5
['Hour' 'Year' 'Week' 'Month'] 4
['Y' 'N'] 2
['Denied' 'Certified'] 2


**advantages**

- using one hot encoder we will get new columns
- the new column value at a time only one will be true ,another will be false.
- which means new columns are independent each other
- independent means no relation
- means no **multi colineraity**
- multi colinearity is animportant property in machine learning
- **ml models expect the column should be independent each other**
- so one hot encoder use the nomulticolineartity

**disadvantages**
- it will increase many columns
- inthe aboveex.total column becomes 25510
- this is happendbecause of case id
- in realtime we will drop the id columns
- eventhough if we drop stillwe are getting 30 columns
- more columns means ,more time to process
- we required more infrastructure(cpu-gpu-tpu)
- this is clled as **curse of dimensionality**